### PR TITLE
add preprocess_string method and use preprocessed loyal trait

### LINKED ISSUE #9580: assign loyal to unit in the editor causes an error
### Game and System Information

- **Version: BfW 1.19.5**
- **OS: Windows 10**

### Description of the bug

When you select **loyal** for a unit in the **scenario editor**, you will cause a scenario inable to be load ( missing macro **OVERLAY**)

### Steps to reproduce the behavior

1. open **editor** -> **new scenario** -> add a **unit** -> click right-> select **loyal**
2. **save** the scenario then **quit** 
3. **load** the scenario again
4. an error is shown and the scenario will be load as a **map** only 

20241113_115307_mfnr
20241113_115329_mfnr

### Expected behavior

**_effect / apply to_**  should be assigned by an image path instead of a macro

### Additional context

the log error is about a missing macro

### PR DESCRIPTION
**Pull request by babaissarkar**
Resolves #9580.

Adds a `preprocess_string` method that allows to run the preprocessor on a string. Example:
https://github.com/wesnoth/wesnoth/blob/0088534cd91b4331fe0ed5e9fcf992e93a1b8ea8/src/editor/map/map_context.cpp#L762

Also, the Loyal trait will now be written inside `[modifications]`.

TODO:
- [x] Add a command line option to preprocess a string

### CODE REVIEW

File: src/serialization/preprocessor.cpp [hunk 1]
```diff
@@ -1747,6 +1747,31 @@ filesystem::scoped_istream preprocess_file(const std::string& fname, preproc_map
...
+	if(!defines) {
+		local_defines.reset(new preproc_map);
+		defines = local_defines.get();
+	}
+
+	buf.reset(new preprocessor_streambuf(defines));
+
+	// Begin processing.
+	buf->add_preprocessor<preprocessor_data>(
+		std::unique_ptr<std::istream>(new std::istringstream(contents)), "", "", 1, "", "wesnoth-lib", nullptr);
```
Comment (soliton-) [inline src/serialization/preprocessor.cpp:1771]: The domain should either be `wesnoth` or given by an argument. Perhaps name should be `<string>` and directory `game_config::path`.

File: src/serialization/preprocessor.hpp [hunk 2]
```diff
@@ -114,6 +114,8 @@ struct preproc_define
 	{
 		return !operator==(v);
 	}
+
+	std::string parsed_value();
```
Comment (soliton-) [inline src/serialization/preprocessor.hpp:118]: This seems unused.

**Review by soliton- — APPROVED** [review]

File: src/serialization/preprocessor.hpp (hunk 2 above)
Comment (babaissarkar) [reply src/serialization/preprocessor.hpp]: removed.

File: src/serialization/preprocessor.cpp (hunk 1 above)
Comment (babaissarkar) [reply src/serialization/preprocessor.cpp]: Used suggeted values.

File: src/commandline_options.hpp [hunk 3]
```diff
@@ -164,6 +164,10 @@ class commandline_options
 	utils::optional<std::string> preprocess_path;
 	/** Target (output) path that was given to the --preprocess option. */
 	utils::optional<std::string> preprocess_target;
+	/** True if --preprocess-string was given on the command line. Starts Wesnoth in preprocessor-only mode. */
+	bool preprocess_string;
```
Comment (soliton-) [inline src/commandline_options.hpp:168]: What's the point of this when `preprocess_source_string` is an optional?

File: src/wesnoth.cpp [hunk 4]
```diff
@@ -123,6 +123,99 @@ static void safe_exit(int res)
...
+		PLAIN_LOG << "Read " << read << " defines.";
+	}
+
+	const std::string source_string(*cmdline_opts.preprocess_source_string);
+	const utils::ms_optimer timer(
+		[](const auto& timer) { PLAIN_LOG << "preprocessing finished. Took " << timer << " ticks."; });
+
+	// If the users add the SKIP_CORE define we won't preprocess data/core
+	bool skipCore = false;
+	bool skipTerrainGFX = false;
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/wesnoth.cpp]: dropped.

File: src/serialization/preprocessor.cpp [hunk 5]
```diff
@@ -1747,6 +1747,31 @@ filesystem::scoped_istream preprocess_file(const std::string& fname, preproc_map
 	return filesystem::scoped_istream(new preprocessor_scope_helper(fname, defines));
 }
 
+std::string preprocess_string(const std::string& contents, preproc_map* defines, const std::string& textdomain)
+{
+	log_scope("preprocessing string " + contents.substr(0, 10) + " ...");
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/serialization/preprocessor.cpp:1752]: logs full string now.

File: src/editor/map/map_context.cpp [hunk 6]
```diff
@@ -756,10 +756,11 @@ config map_context::to_config()
 			u["unrenamable"] = unit.unrenamable();
 		}
 
+		config& mods = u.add_child("modifications");
 		if(unit.loyal()) {
 			config trait_loyal;
-			read(trait_loyal, traits_map["TRAIT_LOYAL"].value);
-			u.append(trait_loyal);
+			read(trait_loyal, preprocess_string("{TRAIT_LOYAL}", &traits_map, "wesnoth-editor"));
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/editor/map/map_context.cpp]: changed to `wesnoth-help` (based on info from `traits.cfg`)

File: src/commandline_options.hpp (hunk 3 above)
Comment (babaissarkar) [reply src/commandline_options.hpp]: dropped.

File: src/wesnoth.cpp [hunk 7]
```diff
@@ -123,6 +123,66 @@ static void safe_exit(int res)
...
+		for(const auto [_, cfg] : cfg.all_children_view()) {
+			const preproc_map::value_type def = preproc_define::read_pair(cfg);
+			defines_map[def.first] = def.second;
+			++read;
+		}
+
+		PLAIN_LOG << "Read " << read << " defines.";
+	}
+
+	const std::string source_string(*cmdline_opts.preprocess_source_string);
```
Comment (soliton-) [inline src/wesnoth.cpp:160]: Is this needed?

File: doc/man/wesnoth.6 [hunk 8]
```diff
@@ -253,24 +253,34 @@ Example:
 For details regarding the preprocessor visit:
 https://wiki.wesnoth.org/PreprocessorRef#Command-line_preprocessor.
 .TP
+.BI --preprocess-string \ source-string
+preprocesses a given string and writes the output to stdout. The common macros
+from the "data/core/macros" directory will be preprocessed before the specified string.
```
Comment (soliton-) [inline doc/man/wesnoth.6:258]: This is now obsolete.

**Review by soliton- — APPROVED** [review]

File: doc/man/wesnoth.6 (hunk 8 above)
Comment (babaissarkar) [reply doc/man/wesnoth.6:258]: Details? What exactly is obsolete, the manpage or something else?

File: src/wesnoth.cpp [hunk 9]
```diff
@@ -123,6 +123,66 @@ static void safe_exit(int res)
...
+	const utils::ms_optimer timer(
+		[](const auto& timer) { PLAIN_LOG << "preprocessing finished. Took " << timer << " ticks."; });
```
[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/wesnoth.cpp:162]: just before L182?

[the remark this answers is not in the document]
Comment (babaissarkar) [reply src/wesnoth.cpp]: Moved.

File: src/wesnoth.cpp (hunk 7 above)
Comment (babaissarkar) [reply src/wesnoth.cpp]: removed.

File: doc/man/wesnoth.6 (hunk 8 above)
Comment (babaissarkar) [reply doc/man/wesnoth.6]: removed.

**Pull request merged**